### PR TITLE
[JENKINS-58054] Add Release Notes support.

### PR DESCRIPTION
**Pull request by eugeniapotapska**
Hi everyone. I have added Release Notes support similar to Hockey App plugin's release notes. Please, look at this PR and write you remarks. I will fix my code ASAP. @mezpahlan

### CODE REVIEW

**Comment by mezpahlan** [comment]
Hi @eugeniapotapska thanks for this contribution and apologies for the delay. I'll start looking at this over the weekend and provide comments. We can work on any changes together on your branch if we need to. Many thanks again.

File: pom.xml [hunk 1]
```diff
@@ -82,7 +82,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins.workflow</groupId>
             <artifactId>workflow-job</artifactId>
-            <scope>test</scope>
+            <version>2.25</version>
```
Comment (mezpahlan) [inline pom.xml]: Is there a good reason why the scope of this has changed from `test`? Same with the other dependencies. They should be `test`.

File: pom.xml [hunk 2]
```diff
@@ -94,7 +94,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins.workflow</groupId>
             <artifactId>workflow-api</artifactId>
-            <scope>test</scope>
+            <version>2.35</version>
```
Comment (mezpahlan) [inline pom.xml]: The version is defined from the BOM which is why it wasn't explicitly defined here. Is there a good reason to add it back it? Defeats the purpose of using the BOM, doesn't it?

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java [hunk 3]
```diff
@@ -53,16 +53,20 @@
...
 
+    @Nonnull
+    private final ReleaseNotes releaseNotesMethod;
+
     @Nullable
     private URL baseUrl;
 
     @DataBoundConstructor
-    public AppCenterRecorder(@Nullable String apiToken, @Nullable String ownerName, @Nullable String appName, @Nullable String distributionGroups, @Nullable String pathToApp) {
+    public AppCenterRecorder(@Nullable String apiToken, @Nullable String ownerName, @Nullable String appName, @Nullable String distributionGroups, @Nullable String pathToApp, ReleaseNotes releaseNotesMethod) {
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: What should the annotation be on `releaseNotesMethod`?

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java [hunk 4]
```diff
@@ -121,7 +128,8 @@ private boolean uploadToAppCenter(Run<?, ?> run, FilePath filePath, TaskListener
                 getOwnerName(),
                 getAppName(),
                 getPathToApp(),
-                getDistributionGroups()
+                getDistributionGroups(),
+                fetchReleaseNotes(run, filePath, run.getEnvironment(taskListener))
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: For consistency this should be `getReleaseNotes()` and it should only return the appropriate release notes method.

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java [hunk 5]
```diff
@@ -131,6 +139,96 @@ private boolean uploadToAppCenter(Run<?, ?> run, FilePath filePath, TaskListener
         }
     }
 
+    private String fetchReleaseNotes(Run<?, ?> build, FilePath filePath, EnvVars vars) {
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: I don't want `AppCenterRecorder` to do a lot of logic. It is much easier to test if it simply configures a job and starts the job. The bulk of the logic should be done away from this class. At the moment that would be in `UploadTask` but in the future I am working on breaking that down in to independent tasks. Either way this logic needs to be moved away from this class.

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java [hunk 6]
```diff
@@ -131,6 +139,96 @@ private boolean uploadToAppCenter(Run<?, ?> run, FilePath filePath, TaskListener
...
+        File tempDir;
+        try {
+            tempDir = File.createTempFile("jtf", null);
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: Do we need to be creating this file handler for _all_ the release notes methods? As far as I can tell it is only for the file release method. The others don't need it Correct?

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java [hunk 7]
```diff
@@ -131,6 +139,96 @@ private boolean uploadToAppCenter(Run<?, ?> run, FilePath filePath, TaskListener
...
+            if (tempDir.delete() && tempDir.mkdirs()) {
+                if (releaseNotesMethod instanceof ManualReleaseNotes) {
+                    ManualReleaseNotes manualReleaseNotes = (ManualReleaseNotes) releaseNotesMethod;
+                    if (((ManualReleaseNotes) releaseNotesMethod).getReleaseNotes() != null) {
+                        return manualReleaseNotes.getReleaseNotes();
+                    }
+                } else if (releaseNotesMethod instanceof FileReleaseNotes) {
+                    FileReleaseNotes fileReleaseNotes = (FileReleaseNotes) releaseNotesMethod;
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: I would also break up this function into smaller ones. Ones that specifically handle the appropriate release notes methods. In fact you could move that logic to the Release notes methods class and return the output whenever you call `getReleaseNotes()`.

File: src/main/java/io/jenkins/plugins/appcenter/releasenotes/ReleaseNotes.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package io.jenkins.plugins.appcenter.releasenotes;
+
+import hudson.ExtensionPoint;
+import hudson.model.Describable;
+
+public abstract class ReleaseNotes implements ExtensionPoint, Describable<ReleaseNotes> { }
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/releasenotes/ReleaseNotes.java]: What's the point of this class? Is it just to identify a type? Why not use an interface instead?

File: src/main/java/io/jenkins/plugins/appcenter/utils/Utils.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package io.jenkins.plugins.appcenter.utils;
+
+import hudson.model.Describable;
+import hudson.model.Descriptor;
+import io.jenkins.plugins.appcenter.Messages;
+import jenkins.model.Jenkins;
+
+public class Utils {
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/utils/Utils.java]: To be honest, I don't really understand why need this file or these helper methods. Was there a use case you had in mind?

File: src/main/resources/io/jenkins/plugins/appcenter/Messages.properties [hunk 10]
```diff
@@ -9,3 +9,11 @@ AppCenterRecorder.DescriptorImpl.errors.invalidDistributionGroups=Distribution G
 AppCenterRecorder.DescriptorImpl.errors.missingPathToApp=Please specify a path to the app you wish to upload
 AppCenterRecorder.DescriptorImpl.errors.invalidPathToApp=Path must be relative
 AppCenterRecorder.DescriptorImpl.DisplayName=Upload app to AppCenter
+Global.errors.JenkinsMissingInstance=Jenkins instance is missing
```
Comment (mezpahlan) [inline src/main/resources/io/jenkins/plugins/appcenter/Messages.properties]: Again, not sure why we need this.

File: src/main/resources/io/jenkins/plugins/appcenter/releasenotes/ManualReleaseNotes/help-releaseNotes.html [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<div>
+    Release notes to be shown to users on App Center.<br/>
+    You can include build variables here like <code>${BUILD_NUMBER}</code> and they'll be expanded at build time.
```
Comment (mezpahlan) [inline src/main/resources/io/jenkins/plugins/appcenter/releasenotes/ManualReleaseNotes/help-releaseNotes.html]: Are you _sure_ they'll be expanded? I have yet to include support for build variables.

File: src/test/java/io/jenkins/plugins/appcenter/FreestyleTest.java [hunk 12]
```diff
@@ -8,6 +8,7 @@
 import hudson.model.Node;
 import hudson.model.Result;
 import hudson.slaves.RetentionStrategy;
+import io.jenkins.plugins.appcenter.releasenotes.NoReleaseNotes;
```
Comment (mezpahlan) [inline src/test/java/io/jenkins/plugins/appcenter/FreestyleTest.java]: I think along with the release notes you should provide tests for the various release notes types and assert that the correct JSON is received by the mockwebserver.

Comment (eugeniapotapska) [reply src/test/java/io/jenkins/plugins/appcenter/FreestyleTest.java]: Yeah, thanks for your review. I will make a new PR with all these remarks ASAP.

**Comment by thewolfwillcome** [comment]
@eugeniapotapska: Any news about a reworked PR for this feature?

**Comment by mezpahlan** [comment]
Hi @thewolfwillcome thanks for chasing this. I am doing some structural changes to the plugin which should make it easier to develop and test new features. So haven't had time to look at this just yet. It is next on my radar.

I'm also in two minds about copying over the HockeyApp code into this plugin for the reasons I've mentioned in the PR comments. My starting position is, "it isn't the responsibility of this plugin to _calculate_ the release notes". We can upload them along with a release by all means, but it's the calculation bit I am not happy with. It just creates problems for me for long term maintenance so I want to keep the plugin more focussed than the old HockeyApp one.

Obviously this causes a bit of a problem for users of the old HockeyApp plugin but with some documentation I think we can overcome that.

Anyway..... keep bugging me I will hope to have some time to look at this again over the weekend.

**Comment by thewolfwillcome** [comment]
Hi @mezpahlan what do you mean with "calculate the release notes". Includes "calculation" expanding of jenkins/env variables?
for my usecase it would be enough to have a property where we could simply put a release note as string which then gets uploaded with the release. But it should at least expand jenkins/env variables (e.g. the svn revision number)

**Comment by mezpahlan** [comment]
Hi @thewolfwillcome yes that's exactly what I'm thinking of. A simply text field that accepts text and optionally expands and environment variables. 

This PR goes one step further and attempts to calculate the release notes based on git information or reading from a file and various things that the old HockeyApp plugin supported.

From my point of view I want to only support a string + environment variable expansion.

**Comment by mezpahlan** [comment]
Hi I just want to update everyone on the state of this. My efforts to refactor the plugin internally are progressing well. I'm looking to publish those first and then support this issue. I don't have an ETA at the moment but sometime this week looks good so far..... work permitting.

**Comment by mezpahlan** [comment]
I've merged #20 now so am ready to have a look at this PR again. Thank you for your patience.

**Comment by carllindberg** [comment]
Would just a string with environment replacements would be flexible enough?  I know that my projects currently use the file method -- we generate release notes during the build to a file, then use that generated file with the current Hockey plugin.  I would imagine some projects have hand-edited release notes files as well, possibly as a checked-in resource on the project, and they may want to use those.

I could maybe understand not having auto-calculation code, at least for initial support (and see how many people have been depending on that), but is the file version that much extra code to add?  Seems like that could add flexibility, and better compatibility with existing setups.

I suppose the entire file could be injected into an environment variable, though sometimes the release notes can get rather long and not sure those are best in the environment (and being logged by other steps, etc.)

**Comment by mezpahlan** [comment]
Hi @carllindberg my thought process is this:
_The existing Hockeyapp plugin is unmaintainable. The code is really hard to reason and has so many overloaded responsibilities that I don't want to simply replicate it here in with the AppCenter plugin._ 

Now, having said that, I think I am coming around to providing something similar in terms of supporting a file based upload for release notes. However as a first step I think I only want to focus on a simple optional parameter that accepts a `String` and understands environment variables for substitution.

**Comment by mezpahlan** [comment]
I've implemented a simple text field that you can use with environment variables for release notes. I am going to close this PR in favour of that one.

I suggest if people want a file upload type release note they can raise a request in Jenkins JIRA and I'll look at it after the other capabilities I want to implement before the HockeyApp shut down.

For now I hope the simple solution will suffice.

**Pull request closed without merging**